homeassistant/components/here_weather/__init__.py

```diff
@@ -0,0 +1,187 @@
+        return False
+    hass.data.setdefault(DOMAIN, {})[config_entry.entry_id] = here_weather_data
+
+    known_api_keys = hass.data.setdefault(HERE_API_KEYS, [])
+    if config_entry.data[CONF_API_KEY] not in known_api_keys:
+        known_api_keys.append(config_entry.data[CONF_API_KEY])
+
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, platform)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN][config_entry.entry_id].unsub_handler()
+        hass.data[DOMAIN].pop(config_entry.entry_id)
+
+    return unload_ok
+
+
+class HEREWeatherData:
+    """Get the latest data from HERE."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize the data object."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self.here_client = herepy.DestinationWeatherApi(config_entry.data[CONF_API_KEY])
+        self.latitude = config_entry.data.get(CONF_LATITUDE)
+        self.longitude = config_entry.data.get(CONF_LONGITUDE)
+        self.location_name = config_entry.data.get(CONF_LOCATION_NAME)
+        self.zip_code = config_entry.data.get(CONF_ZIP_CODE)
+        self.weather_product_type = herepy.WeatherProductType[
+            config_entry.data[CONF_MODE]
+        ]
+        self.units = config_entry.data.get(CONF_UNIT_SYSTEM)
+        self.coordinator = None
+        self.unsub_handler = None
+
+    async def async_setup(self):
+        """Set up the here_weather integration."""
+        self.add_options()
+        self.unsub_handler = self.config_entry.add_update_listener(
+            self.async_options_updated
+        )
+        await self._async_create_coordinator()
+        return True
+
+    def add_options(self):
+        """Add options for here_weather integration."""
+        if not self.config_entry.options:
+            options = {CONF_SCAN_INTERVAL: DEFAULT_SCAN_INTERVAL}
+            self.hass.config_entries.async_update_entry(
+                self.config_entry, options=options
+            )
+
+    async def _async_create_coordinator(self):
+        """Create or recreate the DataUpdateCoordinator."""
+        self.coordinator = DataUpdateCoordinator(
+            self.hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=self.async_update,
+            update_interval=timedelta(
+                seconds=self.config_entry.options[CONF_SCAN_INTERVAL]
+            ),
+        )
+        await self.coordinator.async_refresh()
+
+    async def async_update(self) -> None:
```

Comment:
`async_update()` suggests data is being updated but in fact data is being retrieved. You should consider renaming this method to `_async_get_data_with_timeout` and create `async_update()` which simply calls `self.coordinator.async_refresh()`.
Also, the return type is None but something is being returned.

Reply:
The return type should be list, thank you.

As for the other point I don't understand what you mean. This method is being called by the coordinator not from the outside. It is being supplied in the coordinator constructor as the parameter named "update_method".

Reply:
```suggestion
    async def async_update(self):
```